### ISSUE TITLE
can SVG container elements establish containing block for absolute/fixed positioned elements?

### ISSUE THREAD

**Issue by dbaron** [w3c/fxtf-drafts#307]
Currently the [spec for `filter`](https://drafts.fxtf.org/filter-effects-1/#FilterProperty), if interpreted literally, says that when `filter` is applied to SVG container elements, then those elements establish a containing block for fixed positioned and absolutely positioned elements.

It's not clear to me what this means, since it's not clear to me where the edges of such a containing block are.

On the other hand, it also doesn't make sense for these elements to escape the filter.  One way to avoid that would be to say that `<foreignObject>` establishes such a containing block, but I'm not sure if that would be web compatible.

I think this also applies to a number of other properties (`will-change`, `contain`, and a bunch of transform properties although those properties currently have [an exception](https://drafts.csswg.org/css-transforms/#transform-rendering) for elements not governed by the CSS rendering model... which in turn raises the question of how a transformed element that is position:fixed to an ancestor behaves.

**Comment by tabatkins** [comment]
The proper resolution to this boils down to "actually define how SVG operates in the CSS box model finally", correct? (Aka the thing I've been saying I'd do for a few years now...)

**Comment by AmeliaBR** [comment]
>  One way to avoid that would be to say that <foreignObject> establishes such a containing block, but I'm not sure if that would be web compatible.

That is probably the most logical approach, and most useful. 

The SVG 2 specs define [`<foreignObject>` as creating a stacking context](https://svgwg.org/svg2-draft/render.html#EstablishingStackingContex), so that already covers containing absolutely-positioned elements.  I don't think there is any argument against that.

Fixed positioning is a little more complicated, since currently filters & transforms are the only things that contain fixed elements, and I don't think there's any spec that says SVG should have an effect. But nonetheless, browsers seem to agree that it does, although they don't agree in all the details.

I'm using [this Codepen](https://codepen.io/AmeliaBR/pen/50bbeed851a7c2009ec4b9ddb0c76c4f), which is a modified version of [this WPT test](https://github.com/web-platform-tests/wpt/blob/master/svg/extensibility/foreignObject/containing-block.html).  The WPT test is very problematic; the [failures](https://wpt.fyi/results/svg/extensibility/foreignObject/containing-block.html) are about how `offsetLeft` is calculated, while the passes (Chrome vs WebKit) don't actually agree on layout!

Anyway, here are my results for latest stable Chrome, Edge, Safari & Firefox Dev edition:

- Static (Spring green) and relative (Rebecca Purple)  elements:
  * all browsers position as expected, except that Safari collapses the margins through the edge of the foreignObject container, while the others contain margins.

-  Absolute (Alice Blue) element:
   * Chrome, Firefox & Edge all position relative to the foreignObject parent.
   * Safari positions relative to the initial containing block, I think.

- Fixed (Firebrick) element:
   * All 4 browsers position the element relative to the foreignObject bounds
   * _However_, Chrome then fixes its position relative to scroll on the body, although it disappears if you scroll enough so that is completely outside of the foreignObject bounds.

- Sticky (Steel Blue) element:
   * MS Edge crashes the tab (the element is commented out in the pen to avoid this)
   * Firefox and Chrome position it as expected from static layout, then scrolls it stickily (relative to scroll on the body) while containing it within the foreignObject bounds.
   * Safari doesn't support sticky positioning (I didn't try with `-webkit-`).

So, in conclusion:

I'd say your proposal isn't a web compat problem, but we need some specs & tests to match, and then some bugs filed on browsers.

> The proper resolution to this boils down to "actually define how SVG operates in the CSS box model finally", correct?

That would cover a lot of it, yes...

**Comment by atanassov** [comment]
Establishing stacking context doesn't mean establishing a containing block. Ideally we can define <foreignObject> as establishing an _initial containing block_.

**Comment by chrishtr** [comment]
<foreignObject> is treated as a containing block for all descendants in Chromium. It's also a stacking
context, as of Chromium 67.

The SVG2 spec says, at least indirectly, that <foreignObject> is both a stacking context and containing block. This sentence in particular:

"Since the ‘foreignObject’ element creates a "fixed position containing block" in CSS terms..."

Therefore it's web compatible, as far as it concerns Chromium, to define SVG container elements
as establishing a containing block for all descendants.

**Comment by chrishtr** [comment]
> Fixed (Firebrick) element:
> 
> All 4 browsers position the element relative to the foreignObject bounds
> However, Chrome then fixes its position relative to scroll on the body, although it disappears if you scroll enough so that is completely outside of the foreignObject bounds.

I don't see this behavior. On what version of Chrome are you testing? Could you please file a bug
at crbug.com with exact repro steps?

**Comment by dbaron** [comment]
> Therefore it's web compatible, as far as it concerns Chromium, to define SVG container elements
as establishing a containing block for all descendants.

I'd much rather do this for `<foreignObject>` (as you say Chromium does) but not for SVG container elements in general (both since I think doing it for `<foreignObject>` makes it irrelevant, and because it would be hard to define reasonably).

**Comment by dbaron** [comment]
... and we still need to follow up about whether `position: sticky` requires that we say something about `overflow` or similar, and also whether `<foreignObject>` should be an "*initial* containing block".
